examples/toy/Ch4/mlir/ToyCombine.cpp

```diff
@@ -21,135 +21,78 @@
       auto newAttr = valueAttr.reshape(newType);
-      rewriter.replaceOpWithNewOp<ConstantOp>(reshape, reshapeType.getShape(),
-                                              newAttr);
+      return builder2.create<ConstantOp>(reshape.getLoc(), newType, newAttr);
     } else if (auto valueAttr =
                    constantOp.getAttrOfType<mlir::FloatAttr>("value")) {
       // Broadcast
       auto dataSize = std::accumulate(reshapeType.getShape().begin(),
                                       reshapeType.getShape().end(), 1,
                                       std::multiplies<int>());
       std::vector<mlir::Attribute> data(dataSize, valueAttr);
-      auto tensorTy = rewriter.getTensorType(reshapeType.getShape(),
+      auto tensorTy = builder.getTensorType(reshapeType.getShape(),
                                              reshapeType.getElementType());
       auto newAttr = mlir::DenseElementsAttr::get(tensorTy, data);
-      rewriter.replaceOpWithNewOp<ConstantOp>(reshape, reshapeType.getShape(),
-                                              newAttr);
+      return builder2.create<ConstantOp>(reshape.getLoc(), tensorTy, newAttr);
     } else {
       llvm_unreachable("Unsupported Constant format");
     }
-    return matchSuccess();
-  }
-};
+    return reshape;
+}
 
-/// Fold reshape(reshape(x)) -> reshape(x)
-struct SimplifyReshapeReshape : public mlir::OpRewritePattern<ReshapeOp> {
-  using mlir::OpRewritePattern<ReshapeOp>::OpRewritePattern;
+#include "ToyCombine.inc"
 
-  mlir::PatternMatchResult
-  matchAndRewrite(ReshapeOp op,
-                  mlir::PatternRewriter &rewriter) const override {
-    // Look through the input of the current reshape.
-    mlir::Value *reshapeInput = op.getOperand();
+namespace {
+struct ReshapeOpt : public FunctionPass<ReshapeOpt> {
+  void runOnFunction() override {
+    mlir::OwningRewritePatternList patterns;
+    populateWithGenerated(&getContext(), &patterns);
 
-    // If the input is defined by another reshape, bingo!
-    if (!matchPattern(reshapeInput, mlir::m_Op<ReshapeOp>()))
-      return matchFailure();
+    patterns.insert<ReshapeReshapeOptPattern>(&getContext());
+    patterns.insert<RedundantReshapeOptPattern>(&getContext());
+    patterns.insert<FoldConstantReshapeOptPattern>(&getContext());
 
-    // Use the rewriter to perform the replacement
-    rewriter.replaceOp(op, {reshapeInput});
-    return matchSuccess();
+    applyPatternsGreedily(getFunction(), patterns);
   }
 };
+struct TransposeOpt : public FunctionPass<TransposeOpt> {
+  void runOnFunction() override {
+    mlir::OwningRewritePatternList patterns;
+    populateWithGenerated(&getContext(), &patterns);
 
-/// Fold reshape(x)) -> x, when input type matches output type
-struct SimplifyNullReshape : public mlir::OpRewritePattern<ReshapeOp> {
-  using mlir::OpRewritePattern<ReshapeOp>::OpRewritePattern;
+    patterns.insert<TransposeOptPattern>(&getContext());
```

Comment:
What is the reason to have two passes here? We eventually want all of these passes to go away when we register these patterns as canonicalization patterns on the ops? Actually can you just move these to canonicalization patterns now, given that you are naming the patterns?

Reply:
> Actually can you just move these to canonicalization patterns now
I'm not sure I know how to do this. Now that the ops are defined using ODS, getCanonicalizationPatterns is no longer a member function of the ops.

Reply:
I think you just need to add 'let hasCanonicalizer = 1` to the op definitions:
https://github.com/tensorflow/mlir/blob/master/g3doc/OpDefinitions.md#hascanonicalizer

Reply:
Then you can add the following to the .cpp file.

void *Op::getCanonicalizationPatterns(...) {
}

Reply:
Ah I see. Thanks.